agent/consul/watch/server_local.go

```diff
@@ -0,0 +1,277 @@
+// best-effort attempt to follow these guidelines:
+//
+//   1. Only return an index greater than the minIndex.
+//   2. Any channels added to the memdb.WatchSet should only unblock when the
+//      results returned by the query have changed. This might be difficult
+//      to do when blocking on non-existent data.
+//
+func ServerLocalBlockingQuery[ResultType any, StoreType StateStore](
+	ctx context.Context,
+	getStore func() StoreType,
+	minIndex uint64,
+	suppressSpuriousWakeup bool,
+	query func(memdb.WatchSet, StoreType) (uint64, ResultType, error),
+) (uint64, ResultType, error) {
+	var (
+		notFound  bool
+		ranOnce   bool
+		priorHash uint64
+	)
+
+	var zeroResult ResultType
+	if getStore == nil {
+		return 0, zeroResult, fmt.Errorf("no getStore function was provided to ServerLocalBlockingQuery")
+	}
+	if query == nil {
+		return 0, zeroResult, fmt.Errorf("no query function was provided to ServerLocalBlockingQuery")
+	}
+
+	for {
+		state := getStore()
+
+		ws := memdb.NewWatchSet()
+		ws.Add(state.AbandonCh())
+
+		index, result, err := query(ws, state)
+
+		switch {
+		case errors.Is(err, ErrorNotFound):
+			// if minIndex is 0 then we should never block but we
+			// also should not propagate the error
+			if minIndex == 0 {
+				return index, result, nil
+			}
+
+			// update the min index if the previous result was not found. This
+			// is an attempt to not return data unnecessarily when we end up
+			// watching the root of a memdb Radix tree because the data being
+			// watched doesn't exist yet.
+			if notFound {
+				minIndex = index
+			}
+
+			notFound = true
+		case err != nil:
+			return index, result, err
+		}
+
+		// when enabled we can prevent sending back data that hasn't changed.
+		if suppressSpuriousWakeup {
+			newHash, err := hashstructure_v2.Hash(result, hashstructure_v2.FormatV2, nil)
+			if err != nil {
+				return index, result, fmt.Errorf("error hashing data for spurious wakeup suppression: %w", err)
+			}
+
+			// set minIndex to the returned index to prevent sending back identical data
+			if ranOnce && priorHash == newHash {
+				minIndex = index
+			}
+			ranOnce = true
+			priorHash = newHash
+		}
+
+		// one final check if we should be considered unblocked and
+		// return the value. Some conditions in the switch above
+		// alter the minIndex and prevent this return if it would
+		// be desirable. One such case is when the actual data has
+		// not changed since the last round through the query and
+		// we would rather not do any further processing for unchanged
+		// data. This mostly protects against watches for data that
+		// doesn't exist from return the non-existant value constantly.
+		if index > minIndex {
+			return index, result, nil
+		}
+
+		// block until something changes
+		if err := ws.WatchCtx(ctx); err != nil {
+			// exit if the context was cancelled
+			return index, result, nil
+		}
+
+		select {
+		case <-state.AbandonCh():
+			return index, result, nil
+		default:
+		}
```

Comment:
The first time I read this, I thought that abandoning the state store would cause the above `WatchCtx` to block forever (so we'd never reach this `select`). I now realise that it's fine because we add the `AbandonCh` to the watchset on L106. Would you mind adding a comment to point this out for future readers? 😄

Reply:
Sure. I will add a comment both where we call WatchCtx and where the abandon chan is added to the watch set